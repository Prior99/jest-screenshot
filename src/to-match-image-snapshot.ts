import { decode, readPngFileSync, writePngFileSync, PngImage } from "node-libpng";
import { diffImages, DiffImage } from "native-image-diff";
import chalk from "chalk";
import { existsSync, writeFileSync, readFileSync } from "fs";
import { getSnapshotPath, getReportPath, getReportDir } from "./filenames";
import { isJestTestConfiguration, MatcherResult } from "./jest";
import { sync as mkdirp } from "mkdirp";
import * as path from "path";
import { JestScreenshotConfiguration } from "./config";

export interface ImageMatcherResult extends MatcherResult {
    diffImage?: DiffImage;
    changedRelative?: number;
    totalPixels?: number;
    changedPixels?: number;
    testFileName?: string;
    snapshotNumber?: number;
}

export interface ToMatchImageSnapshotParameters {
    /**
     * Can be used to override the path to which the snapshot image
     * will be written.
     */
    path?: string;
}

/**
 * Performs the actual check for equality of two images.
 *
 * @param snapshotImage The image from the snapshot.
 * @param receivedImage The image received from the `expect(...)` call.
 * @param snapshotNumber The number of the snapshot in this test.
 * @param configuration The configuration of the call to `toMatchImageSnapshot`.
 *
 * @return A `MatcherResult` with `pass` and a message which can be handed to jest.
 */
function checkImages(
    snapshotImage: PngImage,
    receivedImage: PngImage,
    snapshotNumber: number,
    configuration: JestScreenshotConfiguration,
): ImageMatcherResult {
    const {
        colorThreshold,
        detectAntialiasing,
        pixelThresholdAbsolute,
        pixelThresholdRelative,
    } = configuration;
    // Perform the actual image diff.
    const { pixels: changedPixels, image: diffImage } = diffImages({
        image1: receivedImage,
        image2: snapshotImage,
        colorThreshold,
        detectAntialiasing,
    });
    const expected = `stored snapshot ${snapshotNumber}`;
    const preamble = `${chalk.red("Received value")} does not match ${chalk.green(expected)}.`;
    const snapshotImagePixels = snapshotImage.width * snapshotImage.height;
    const receivedImagePixels = receivedImage.width * receivedImage.height;
    const totalPixels = Math.max(snapshotImagePixels, receivedImagePixels);
    const changedRelative = changedPixels / totalPixels;
    if (typeof pixelThresholdAbsolute === "number" && changedPixels > pixelThresholdAbsolute) {
        return {
            pass: false,
            message: () =>
                `${preamble}\n\n` +
                `Expected less than ${chalk.green(`${pixelThresholdAbsolute} pixels`)} to have changed, ` +
                `but ${chalk.red(`${changedPixels} pixels`)} changed.`,
            diffImage,
            changedRelative,
            totalPixels,
            changedPixels,
        };
    }
    if (typeof pixelThresholdRelative === "number" && changedRelative > pixelThresholdRelative) {
        const percentThreshold = (pixelThresholdRelative * 100).toFixed(2);
        const percentChanged = (changedRelative * 100).toFixed(2);
        return {
            pass: false,
            message: () =>
                `${preamble}\n\n` +
                `Expected less than ${chalk.green(`${percentThreshold}%`)} of the pixels to have changed, ` +
                `but ${chalk.red(`${percentChanged}%`)} of the pixels changed.`,
            diffImage,
            changedRelative,
            totalPixels,
            changedPixels,
        };
    }
    return { pass: true };
}

/**
 * A matcher for jest with compares a PNG image to a stored snapshot. Behaves similar to `.toMatchSnapshot()`.
 *
 * @param received The buffer from the call to `expect(...)`.
 * @param configuration The configuration object provided when initializing this library
 *     with a call to `jestScreenshot`.
 * @param parameters Optional parameters provided to the call of `expect(...).toMatchImageSnapshot(...)`.
 *
 * @return A `MatcherResult` usable by jest.
 */
export function toMatchImageSnapshot(
    received: Buffer,
    configuration: JestScreenshotConfiguration,
    parameters: ToMatchImageSnapshotParameters = {},
): MatcherResult {
    const { snapshotsDir, reportDir, noReport } = configuration;
    // Check whether `this` is really the expected Jest configuration.
    if (!isJestTestConfiguration(this)) {
        throw new Error("Jest: Attempted to call `.toMatchImageSnapshot()` outside of Jest context.");
    }
    const { testPath, currentTestName, isNot } = this;
    if (isNot) {
        throw new Error("Jest: `.not` cannot be used with `.toMatchImageSnapshot()`.");
    }
    let { snapshotState } = this;
    const { _updateSnapshot } = snapshotState;
    const snapshotNumber = (snapshotState._counters.get(currentTestName) || 0) as number + 1;
    snapshotState._counters.set(currentTestName, snapshotNumber);
    const snapshotPath = typeof parameters.path === "string" ?
        parameters.path :
        getSnapshotPath(testPath, currentTestName, snapshotState, snapshotsDir);
    const reportPath = getReportPath(testPath, currentTestName, snapshotState, reportDir);
    // Create the path to store the snapshots in.
    mkdirp(path.dirname(snapshotPath));
    // The image did not yet exist.
    if (!existsSync(snapshotPath)) {
        // If the user specified `-u`, or was running in interactive mode, write the new
        // snapshot to disk and let the test pass.
        if (_updateSnapshot === "new" || _updateSnapshot === "all") {
            snapshotState.added++;
            writeFileSync(snapshotPath, received);
            return { pass: true };
        }
        // Otherwise fail due to missing snapshot.
        return {
            pass: false,
            message: () => `New snapshot was ${chalk.red("not written")}. ` +
                `The update flag must be explicitly passed to write a new snapshot.\n\n` +
                `This is likely because this test is run in a continuous integration (CI) environment ` +
                `in which snapshots are not written by default.`,
        };
    }

    // If received and expected Buffers are equal, there is no need to perform actual check
    const snapshotBuffer = readFileSync(snapshotPath);
    if (snapshotBuffer.equals(received)) {
        return { pass: true };
    }

    // Decode the new image and read the snapshot.
    const snapshotImage = decode(snapshotBuffer);
    const receivedImage = decode(received);
    // Perform the actual diff of the images.
    const {
        pass,
        message,
        diffImage,
        changedRelative,
        totalPixels,
        changedPixels,
    } = checkImages(snapshotImage, receivedImage, snapshotNumber, configuration);
    if (!pass) {
        if (_updateSnapshot === "all") {
            snapshotState.updated++;
            writeFileSync(snapshotPath, received);
            return { pass: true };
        }
        if (!noReport) {
            mkdirp(reportPath);
            const receivedPath = path.join(reportPath, "received.png");
            const diffPath = path.join(reportPath, "diff.png");
            const snapshotPathReport = path.join(reportPath, "snapshot.png");
            writeFileSync(receivedPath, received);
            writeFileSync(snapshotPathReport, readFileSync(snapshotPath));
            writePngFileSync(diffPath, diffImage.data, diffImage);
            writeFileSync(path.join(reportPath, "info.json"), JSON.stringify({
                testName: currentTestName,
                message: message(),
                changedRelative,
                totalPixels,
                changedPixels,
                testFileName: path.relative(process.cwd(), testPath),
                snapshotNumber,
                receivedPath: path.relative(getReportDir(reportDir), receivedPath),
                diffPath: path.relative(getReportDir(reportDir), diffPath),
                snapshotPath: path.relative(getReportDir(reportDir), snapshotPathReport),
                width: Math.max(snapshotImage.width, receivedImage.width),
                height: Math.max(snapshotImage.height, receivedImage.height),
            }));
        }
    }
    return { pass, message };
}
