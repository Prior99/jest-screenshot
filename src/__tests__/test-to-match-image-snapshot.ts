import { setupJestScreenshot } from "..";
import { toMatchImageSnapshot } from "../to-match-image-snapshot";
import { SnapshotState, JestTestConfiguration } from "../jest";
import { readdirSync, readFileSync, unlinkSync, existsSync, writeFileSync } from "fs";
import * as rimraf from "rimraf";

function getJestTestConfiguration(): JestTestConfiguration {
    let testConfiguration: JestTestConfiguration;
    expect.extend({
        storeSnapshotState() {
            testConfiguration = this as any;
            return {
                pass: true,
                message() { return ""; },
            };
        },
    });
    (expect as any)().storeSnapshotState();
    return testConfiguration;
}

describe("toMatchImageSnapshot", () => {
    let testConfig: JestTestConfiguration;
    let originalUpdateSnapshot: "new" | "all" | "none";

    beforeEach(() => {
        setupJestScreenshot();
    });

    beforeEach(() => {
        testConfig = getJestTestConfiguration();
        originalUpdateSnapshot = testConfig.snapshotState._updateSnapshot;
        try {
            unlinkSync(`${process.cwd()}/jest-screenshot.json`);
        } catch (err) {
            return;
        }
    });

    afterEach(() => {
        testConfig.snapshotState._updateSnapshot = originalUpdateSnapshot;
    });

    describe("with no threshold provided", () => {
        it("detects a matching snapshot as matching", () => {
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-gradient.png`)).toMatchImageSnapshot();
            }).not.toThrowError();
        });

        [
            { colorThreshold: 0.0 },
            { colorThreshold: 0.1 },
            { colorThreshold: 0.5 },
            { colorThreshold: 0.6 },
        ].forEach(({ colorThreshold }) => {
            it(`with a color threshold of ${colorThreshold} detects the snapshot as not matching`, () => {
                writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({ colorThreshold }));
                setupJestScreenshot();
                expect(() => {
                    expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
                }).toThrowErrorMatchingSnapshot();
            });

            it(`with a color threshold of ${colorThreshold} provided through matcher params detects the snapshot as not matching`, () => {
                writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({}));
                setupJestScreenshot();
                expect(() => {
                    expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`))
                      .toMatchImageSnapshot({colorThreshold});
                }).toThrowErrorMatchingSnapshot();
            });
        });

        [
            { colorThreshold: 0.7 },
            { colorThreshold: 1.0 },
        ].forEach(({ colorThreshold }) => {
            it(`with a color threshold of ${colorThreshold} detects the snapshot as matching`, () => {
                writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({ colorThreshold }));
                setupJestScreenshot();
                expect(() => {
                    expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
                }).not.toThrowError();
            });

            it(`with a color threshold of ${colorThreshold} provided through matcher params detects the snapshot as matching`, () => {
                writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({  }));
                setupJestScreenshot();
                expect(() => {
                    expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`))
                      .toMatchImageSnapshot({colorThreshold});
                }).not.toThrowError();
            });
        });
    });

    describe("with an absolute threshold provided", () => {
        const colorThreshold = 0.6;

        it("fails with an absolute threshold of 204", () => {
            const pixelThresholdAbsolute = 204;
            writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({
                colorThreshold,
                pixelThresholdAbsolute,
            }));
            setupJestScreenshot();
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).toThrowErrorMatchingSnapshot();
        });

        it("fails with an absolute threshold of 204 passed through matcher params", () => {
            const pixelThresholdAbsolute = 204;
            writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({
                colorThreshold,
            }));
            setupJestScreenshot();
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`))
                  .toMatchImageSnapshot({pixelThresholdAbsolute});
            }).toThrowErrorMatchingSnapshot();
        });

        it("passes with an absolute threshold of 205", () => {
            const pixelThresholdAbsolute = 205;
            writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({
                colorThreshold,
                pixelThresholdAbsolute,
            }));
            setupJestScreenshot();
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).not.toThrowError();
        });

        it("passes with an absolute threshold of 205 passed through matcher params", () => {
            const pixelThresholdAbsolute = 205;
            writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({
                colorThreshold,
            }));
            setupJestScreenshot();
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`))
                  .toMatchImageSnapshot({
                      pixelThresholdAbsolute,
                  });
            }).not.toThrowError();
        });
    });

    describe("with `.not` specified", () => {
        it("fails always", () => {
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).not.toMatchImageSnapshot();
            }).toThrowErrorMatchingSnapshot();
        });
    });

    describe("when called outside of a jest unit test context", () => {
        it("fails", () => {
            expect(function() {
                toMatchImageSnapshot(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`), {});
            }.bind({})).toThrowErrorMatchingSnapshot();
        });
    });

    describe("with the snapshot not matching", () => {
        const snapshotToUpdatePath = `${__dirname}/__snapshots__/test-to-match-image-snapshot-ts-to-match-image-snapshot-with-the-snapshot-not-matching-updates-the-snapshot-when-updating-is-enabled-1-d8503.snap.png`; // tslint:disable-line
        const originalContent = readFileSync(snapshotToUpdatePath);

        afterEach(() => {
            try {
                writeFileSync(snapshotToUpdatePath, originalContent);
            } catch (err) {
                return;
            }
        });

        it("updates the snapshot when updating is enabled", () => {
            const expectedContent = readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`);
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "all";
                expect(expectedContent).toMatchImageSnapshot();
            }).not.toThrowError();
            expect(readFileSync(snapshotToUpdatePath)).toEqual(expectedContent);
        });

        it("fails when updating is disabled", () => {
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "none";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).toThrowErrorMatchingSnapshot();
            expect(readFileSync(snapshotToUpdatePath)).toEqual(originalContent);
        });
    });

    describe("with the snapshot not existing", () => {
        const snapshotToCreatePath1 = `${__dirname}/__snapshots__/test-to-match-image-snapshot-ts-to-match-image-snapshot-with-the-snapshot-not-existing-creates-the-snapshot-when-updating-is-enabled-1-14292.snap.png`; // tslint:disable-line
        const snapshotToCreatePath2 = `${__dirname}/__snapshots__/test-to-match-image-snapshot-ts-to-match-image-snapshot-with-the-snapshot-not-existing-creates-the-snapshot-when-updating-is-enabled-2-078e4.snap.png`; // tslint:disable-line

        afterEach(() => {
            try {
                unlinkSync(snapshotToCreatePath1);
                unlinkSync(snapshotToCreatePath2);
            } catch (err) {
                return;
            }
        });

        it("fails when updating is disabled", () => {
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "none";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).toThrowErrorMatchingSnapshot();
        });

        it("creates the snapshot when updating is enabled", () => {
            expect(existsSync(snapshotToCreatePath1)).toBe(false);
            expect(existsSync(snapshotToCreatePath2)).toBe(false);
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "all";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).not.toThrowError();
            expect(existsSync(snapshotToCreatePath1)).toBe(true);
            expect(existsSync(snapshotToCreatePath2)).toBe(true);
        });
    });

    it("after a failing snapshot test creates the necessary reports", () => {
        rimraf.sync(`${process.cwd()}/jest-screenshot-report`);

        setupJestScreenshot();
        expect(() => {
            expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
        }).toThrowError();

        const snapshotFilename = "test-to-match-image-snapshot-ts-to-match-image-snapshot-after-a-failing-snapshot-test-creates-the-necessary-reports-1-86dd5.snap.png"; // tslint:disable-line
        const contents = readdirSync(`${process.cwd()}/jest-screenshot-report/reports`);
        expect(contents).toContain(snapshotFilename);
        const snapshotContents = readdirSync(`${process.cwd()}/jest-screenshot-report/reports/${snapshotFilename}`);
        expect(snapshotContents).toEqual([
            "diff.png",
            "info.json",
            "received.png",
            "snapshot.png",
        ]);
        const infoFileContents =
            readFileSync(`${process.cwd()}/jest-screenshot-report/reports/${snapshotFilename}/info.json`, "utf8");
        expect(JSON.parse(infoFileContents)).toMatchSnapshot();
    });

    it("doesn't create a report with `noReport` set to `true`", () => {
        rimraf.sync(`${process.cwd()}/jest-screenshot-report`);

        writeFileSync(`${process.cwd()}/jest-screenshot.json`, JSON.stringify({ noReport: true }));
        setupJestScreenshot();
        expect(() => {
            expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
        }).toThrowError();

        expect(existsSync(`${process.cwd()}/jest-screenshot-report`)).toBe(false);
        unlinkSync(`${process.cwd()}/jest-screenshot.json`);
    });

    describe("with a custom path provided", () => {
        const path = `${__dirname}/custom-path.png`;
        beforeAll(() => {
            rimraf.sync(`${process.cwd()}/jest-screenshot-report`);
            setupJestScreenshot();
        });

        afterAll(() => {
            unlinkSync(path);
        });

        it("writes a new screenshot", () => {
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "all";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-gradient.png`)).toMatchImageSnapshot({
                    path,
                });
            }).not.toThrowError();
            expect(existsSync(path)).toBe(true);
        });

        it("fails with a non-matching screenshot", () => {
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "none";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot({
                    path,
                });
            }).toThrowErrorMatchingSnapshot();
        });

        it("fails with a non-matching screenshot and writes the report", () => {
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "none";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot({
                    path,
                });
            }).toThrowErrorMatchingSnapshot();
            const snapshotFilename = "test-to-match-image-snapshot-ts-to-match-image-snapshot-with-a-custom-path-provided-fails-with-a-non-matching-screenshot-and-writes-the-report-1-e3f07.snap.png"; // tslint:disable-line
            const contents = readdirSync(`${process.cwd()}/jest-screenshot-report/reports`);
            expect(contents).toContain(snapshotFilename);
        });

        it("updates the snapshot when updating is enabled", () => {
            const expectedContent = readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`);
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "all";
                expect(expectedContent).toMatchImageSnapshot({ path });
            }).not.toThrowError();
            expect(readFileSync(path)).toEqual(expectedContent);
        });
    });
});
