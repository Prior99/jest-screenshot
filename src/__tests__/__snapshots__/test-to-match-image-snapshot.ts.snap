// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchImageSnapshot after a failing snapshot test creates the necessary reports 2`] = `
Object {
  "changedPixels": 2272,
  "changedRelative": 0.4544,
  "diffPath": "reports/test-to-match-image-snapshot-ts-to-match-image-snapshot-after-a-failing-snapshot-test-creates-the-necessary-reports-1-86dd5.snap.png/diff.png",
  "height": 50,
  "message": "[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m45.44%[39m of the pixels changed.",
  "receivedPath": "reports/test-to-match-image-snapshot-ts-to-match-image-snapshot-after-a-failing-snapshot-test-creates-the-necessary-reports-1-86dd5.snap.png/received.png",
  "snapshotNumber": 1,
  "snapshotPath": "reports/test-to-match-image-snapshot-ts-to-match-image-snapshot-after-a-failing-snapshot-test-creates-the-necessary-reports-1-86dd5.snap.png/snapshot.png",
  "testFileName": "src/__tests__/test-to-match-image-snapshot.ts",
  "testName": "toMatchImageSnapshot after a failing snapshot test creates the necessary reports",
  "totalPixels": 5000,
  "width": 100,
}
`;

exports[`toMatchImageSnapshot when called outside of a jest unit test context fails 1`] = `"Jest: Attempted to call \`.toMatchImageSnapshot()\` outside of Jest context."`;

exports[`toMatchImageSnapshot with \`.not\` specified fails always 1`] = `"Jest: \`.not\` cannot be used with \`.toMatchImageSnapshot()\`."`;

exports[`toMatchImageSnapshot with a custom path provided fails with a non-matching screenshot 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m45.44%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with a custom path provided fails with a non-matching screenshot and writes the report 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m45.44%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with an absolute threshold provided fails with an absolute threshold of 204 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m204 pixels[39m to have changed, but [31m205 pixels[39m changed."
`;

exports[`toMatchImageSnapshot with an absolute threshold provided fails with an absolute threshold of 204 passed through matcher params 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m204 pixels[39m to have changed, but [31m205 pixels[39m changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0 detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m53.24%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0 provided through matcher params detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m53.24%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.1 detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m45.44%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.1 provided through matcher params detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m45.44%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.5 detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m12.52%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.5 provided through matcher params detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m12.52%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.6 detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m4.10%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.6 provided through matcher params detects the snapshot as not matching 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m4.10%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with the snapshot not existing fails when updating is disabled 2`] = `
"New snapshot was [31mnot written[39m. The update flag must be explicitly passed to write a new snapshot.

This is likely because this test is run in a continuous integration (CI) environment in which snapshots are not written by default."
`;

exports[`toMatchImageSnapshot with the snapshot not matching fails when updating is disabled 2`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m45.44%[39m of the pixels changed."
`;
